aiken/lib/cardano_loans/payment_validation.ak

```diff
@@ -278,12 +279,16 @@ fn process_input(
         )
       )
 
+      let requires_interest = compounding_interest && interest_num > 0
+
       expect True = error_if_false(
         @"Next interest and/or penalty application required",
-        // If the interest is set to zero and penalty is NoPenalty, the compound_frequency is irrelevant.
-        (interest_num == 0 && penalty == NoPenalty) || when compound_frequency is {
+        // If the interest is non-compounding or the interest is set to zero, and penalty is 
+        // NoPenalty, the epoch_duration is irrelevant. No rollover is required in this situation.
+        // Also, no rollover is required when epoch_duration is None.
+        (!requires_interest && penalty == NoPenalty) || when epoch_duration is {
           None -> True
-          Some(freq) -> upper <= last_compounding + freq
+          Some(freq) -> upper <= last_epoch_boundary + freq
         }
       )
```

Comment:
Determining whether interest/penalties must be applied now accounts for whether the interest is compounding or not. The interest must be both compounding and greater than zero for interest to need to be applied.